File: Objective-C/LRI/NetworkLayer/NetworkLayer.m
//
//  NetworkLayer.m
//  LRI
//
//  Created by Development on 05.07.2022.
//

#import <Foundation/Foundation.h>
#import "NetworkLayer.h"
#import "DrinkEndpoint.h"
#import "DrinkItem.h"

@interface NetworkLayer ()

@end

@implementation NetworkLayer

- (NSURLSessionDataTask *)performRequestWithEndpoint:(id<Endpoint>) endpoint withCompletion:(void (^)(NSMutableArray<DrinkItem *> *items, NSError *error)) completion {
    NSURLComponents *components = [[NSURLComponents alloc] init];
    [components setPath: endpoint.path];
    [components setScheme: endpoint.scheme];
    [components setHost: endpoint.baseURL];
    [components setQueryItems: endpoint.parameters];
    
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
    [request setURL:components.URL];
    [request setHTTPMethod: endpoint.method];
    
    NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    
    NSURLSessionDataTask *dataTask = [session dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
 
        NSError *decodingError;
        NSMutableArray *jsonArray = (NSMutableArray *)[NSJSONSerialization JSONObjectWithData:data options:NSASCIIStringEncoding error:&decodingError];
        
        if ((error != nil) || (decodingError != nil)) {
            NSError *customError = [[NSError alloc] initWithDomain:NSCocoaErrorDomain code:200 userInfo:@{
                NSLocalizedDescriptionKey:@"The server responded with an error."}];
            completion(NSMutableArray.new, customError);
            return;
        }
                
        NSMutableArray <DrinkItem *> *items = NSMutableArray.new;
        for (NSMutableDictionary *dictionary in [jsonArray valueForKey:@"drinks"]) {
            DrinkItem *item = [DrinkItem withDictionary:dictionary];
            [items addObject:item];
        }
        
        completion(items, nil);
    }];
    [dataTask resume];
    return dataTask;
}

@end

File: SwiftUIKit/LRI/NetworkLayer/NetworkLayer.swift
//
//  NetworkLayer.swift
//  LRI
//
//  Created by Development on 01.07.2022.
//

import Foundation

struct NetworkLayer {
    @discardableResult
    static func request<T: Codable>(endpoint: Endpoint,
                                    completion: @escaping (Result<T, Error>) -> ())
    -> URLSessionDataTask? {
        var components = URLComponents()
        components.scheme = endpoint.scheme
        components.host = endpoint.baseURL
        components.path = endpoint.path
        components.queryItems = endpoint.parameters
        
        guard let url = components.url else { return nil }
        
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = endpoint.method
        
        let dataTask = URLSession.shared.dataTask(with: urlRequest) { data, response, error in
            guard error == nil else {
                completion(.failure(error!))
                return
            }
            
            guard (response as? HTTPURLResponse)?.statusCode == 200 else {
                let error = NSError(domain: "", code: NSURLErrorBadServerResponse, userInfo: [NSLocalizedDescriptionKey: "The server responded with an error."])
                completion(.failure(error))
                return
            }
            
            guard let data = data else {
                let error = NSError(domain: "", code: NSURLErrorZeroByteResource, userInfo: [NSLocalizedDescriptionKey: "Server terminated the network connection without sending any data."])
                completion(.failure(error))
                return
            }

            guard let decoded = try? JSONDecoder().decode(T.self, from: data) else {
                let error = NSError(domain: "", code: NSURLErrorCannotDecodeContentData, userInfo: [NSLocalizedDescriptionKey: "Failed to decode response."])
                completion(.failure(error))
                return
            }

            completion(.success(decoded))
        }
        dataTask.resume()
        
        return dataTask
    }
}
